### PR TITLE
Add write load to replication and self healing tests

### PR DESCRIPTION
**Pull request by paulomach**
# Issue

Replication and self healing tests were not being executed under load.

# Solution

Include load charm and test fixture to use it.

# Context

There's a great deal of refactor on the test code and file structure (keep same structure as in k8s).
Some opportunistic updates were done and can be safely dismissed from the review, namely:
* CI action versions updates
* upstream charm lib updates (tls, snap)

# Testing

That's what this was all about ;) 

# Release Notes

Review worthy files:
* lib/charms/mysql/v0/mysql.py - added new method to get endpoints in a single db call (was being done in two db calls)
* src/relations/mysql_provider.py - using above and add handler to update endpoints on `update_status` - necessary for self healing scenarios where no event is triggered
* tests/integration/helpers.py - name refactor and new method
* tests/integration/high_availability/application_charm/*  - verbatim copy from k8s test app - can be dismissed
* tests/integration/high_availability/conftest.py - fixture to manage load app (from k8s also)
* tests/integration/high_availability/high_availability_helpers.py - helpers for tests using load - almost verbatim copy from k8s implementation
* tests/integration/high_availability/test_replication.py - replication tests with load
* tests/integration/high_availability/test_self_healing.py - self healing test with load

### CODE REVIEW

**Review by shayancanonical — APPROVED** [review]
Looks great! Just left a minor nit

File: tests/integration/high_availability/test_replication.py [hunk 1]
```diff
@@ -85,131 +54,82 @@ async def test_consistent_data_replication_across_cluster(
...
 
-    random_chars = generate_random_string(40)
-    create_records_sql = [
-        "CREATE DATABASE IF NOT EXISTS test",
-        "CREATE TABLE IF NOT EXISTS test.data_replication_table (id varchar(40), primary key(id))",
-        f"INSERT INTO test.data_replication_table VALUES ('{random_chars}')",
-    ]
+    database_name, table_name = "test-check-consistency", "data"
+    await insert_data_into_mysql_and_validate_replication(ops_test, database_name, table_name)
+    await clean_up_database_and_table(ops_test, database_name, table_name)
```
Comment (marceloneppel) [inline tests/integration/high_availability/test_replication.py:64]: Would it be a good idea to move this to a fixture to cleanup the database and table if the `insert_data_into_mysql_and_validate_replication ` function fails?

**Review by marceloneppel — APPROVED** [review]
Great PR, Paulo!

File: tests/integration/high_availability/application_charm/src/continuous_writes.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+                )
+        except mysql.connector.errors.DatabaseError as e:
+            # errno 2003: can't connect to mysql database
+            if e.errno != 2003:
+                raise
+            continue
+        except Exception as e:
+            if e.errno != 2013 or e.msg != "Lost connection to MySQL server during query":
+                raise
+            continue
```
Comment (MiaAltieri) [inline tests/integration/high_availability/application_charm/src/continuous_writes.py:46]: might be nice to add comments with your MySQL specific knowledge on why you continue vs raise here

File: tests/integration/high_availability/conftest.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+    clear_writes_action = await application_unit.run_action("clear-continuous-writes")
+    await clear_writes_action.wait()
+
+    start_writes_action = await application_unit.run_action("start-continuous-writes")
+    await start_writes_action.wait()
+
+    yield
+
+    clear_writes_action = await application_unit.run_action("clear-continuous-writes")
```
Comment (MiaAltieri) [inline tests/integration/high_availability/conftest.py:25]: I assume clear also stops the continuous writes?

File: src/charm.py [hunk 4]
```diff
@@ -483,12 +484,15 @@ def _workload_reset(self) -> StatusBase:
             A `StatusBase` to be set by the caller
         """
         try:
+            primary_address = self._get_primary_address_from_peers()
+            if not primary_address:
+                logger.debug("Primary not yet defined on peers. Skipping workload reset")
+                return MaintenanceStatus("Workload reset")
```
Comment (MiaAltieri) [inline src/charm.py:490]: aren't these a little conflicting messages? 
Logging that you're skipping reseting the workload - but. then setting maintaince message that you are resetting the workload

File: tests/integration/high_availability/test_self_healing.py [hunk 5]
```diff
@@ -155,22 +140,27 @@ async def test_freeze_db_process(ops_test: OpsTest):
...
-        ops_test, primary_unit.name, another_unit.name
-    ), "❌ Unit not online in the cluster"
+    # ensure continuous writes still incrementing for all units
+    async with ops_test.fast_forward():
+        await ensure_all_units_continuous_writes_incrementing(ops_test)
 
+    # ensure that we are able to insert data into the primary and have it replicated to all units
+    database_name, table_name = "test-freeze-db-process", "data"
+    await insert_data_into_mysql_and_validate_replication(ops_test, database_name, table_name)
+    await clean_up_database_and_table(ops_test, database_name, table_name)
```
Comment (MiaAltieri) [inline tests/integration/high_availability/test_self_healing.py:150]: would be good to put `clean_up_database_and_table ` in a fixture to gaurentee that the database and table are cleaned up after the test

File: tests/integration/high_availability/test_self_healing.py [hunk 6]
```diff
@@ -222,20 +212,29 @@ async def test_network_cut(ops_test: OpsTest):
...
 
-    # verify instance is part of the cluster
-    logger.info("Check if instance in cluster")
-    assert await is_unit_in_cluster(
-        ops_test, primary_unit.name, another_unit.name
-    ), "Unit not online in the cluster"
+    # ensure continuous writes still incrementing for all units
+    async with ops_test.fast_forward():
+        # allow extra time for relation data to be updated, accommodating CI lag
+        sleep(60)
```
Comment (MiaAltieri) [inline tests/integration/high_availability/test_self_healing.py:218]: If this is giving you trouble, then maybe there is a condition that you could check and then you could throw in an `ops.model.wait_for`

**Review by MiaAltieri — APPROVED** [review]
looks good to me

**Pull request merged**